From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kamillaova <54859825+Kamillaova@users.noreply.github.com>
Date: Fri, 25 Nov 2022 00:44:39 +0300
Subject: [PATCH] Prevent obsessive and strange class leaks from other plugins


diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index d2fa792e182736070a119e96b4ae79337de53d2a..e62bc887209362f9398ad3714237a2e086215116 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -139,6 +139,20 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
                         && !seenIllegalAccess.contains(provider.getName())
                         && !((SimplePluginManager) loader.server.getPluginManager()).isTransitiveDepend(description, provider)) {
 
+                    if (
+                      // JDBC Drivers
+                      name.endsWith("org.mariadb.jdbc.Driver") ||
+                      name.endsWith("com.mysql.cj.jdbc.Driver") ||
+                      name.endsWith("org.h2.Driver") ||
+                      name.endsWith("org.hsqldb.jdbc.JDBCDriver") ||
+                      name.endsWith("org.postgresql.Driver") ||
+                      name.endsWith("com.impossibl.postgres.jdbc.PGDriver") ||
+                      // Misc classes
+                      name.startsWith("org.apache.log4j") // 1000iq UltimateClans developer moment // TODO: Remove it
+                    ) {
+                        throw new ClassNotFoundException(name);
+                    }
+
                     seenIllegalAccess.add(provider.getName());
                     if (plugin != null) {
                         plugin.getLogger().log(Level.WARNING, "Loaded class {0} from {1} which is not a depend, softdepend or loadbefore of this plugin.", new Object[]{name, provider.getFullName()});
